### ISSUE TITLE
Framework cannot be embedded in iOS Extension due to use of restricted APIs.

### ISSUE THREAD

**Issue by Antoine4011** [danthorpe/Operations#84]
You're using UIApplication.sharedApplication(), but it's not available for extension. 

Could you please make it optional if it isn't available? :)

Thanks.

**Comment by danthorpe** [comment]
Yeah, ok. Will look into it tomorrow.

**Comment by danthorpe** [comment]
Hi @Antoine4011 I've been looking into this - I've not built an extension before, so was not aware of this restriction. But I'm all over it now 😀.

The `UIApplication().sharedApplication` is used for a couple of features, `BackgroundObserver`, `NetworkObserver` and `RemoteNotificationCondition`. To proceed there are a couple of options:

1. These types are designed (for the tests) to have underlying types injected into them, so I could remove the restricted APIs entirely, and consumers would need to write code like:

    ```swift
    operation.addCondition(BackgroundObserver(UIApplication().sharedApplication))
    ```

    instead of what is currently required

    ```swift
    operation.addCondition(BackgroundObserver())
    ```

2. I think it's safe to say that these features (background tasks, remote notifications etc) are also not allowed in extensions. Therefore, another option would be to have a "core" framework, which is suitable for Extensions, and then an "application" framework, suitable for applications. For Cocoapods, this could be done with sub-specs.

I'm keen to hear from a developer who is actively working on extensions and shared frameworks on what they think is best. What do you think? I'm leaning towards option 2 myself.

Can you tell me a little bit about how you would integrate Operations into your project?

**Comment by Antoine4011** [comment]
Hi @danthorpe 

For my project, I would like to integrate it via pods in a shared framework for all my targets (iOS, watchOS, tvOS) so your second solution works I think. MixPanel chose this solution https://github.com/mixpanel/mixpanel-iphone/releases/tag/v2.8.0

A third option would be to add a preprocessor macro and to disable the unavailable code for extensions target. AFNetworking chose this solution https://github.com/AFNetworking/AFNetworking/issues/2119

Maybe the third option is the easier, even though I think the second one is cleaner. 

So I guess it's up to you to choose what's best 😊

Thanks!

**Comment by danthorpe** [comment]
Okay, I'm gonna do option 2, as it will also scale to OS X targets too. Stay tuned, should get this done Monday or Tuesday of next week.

**Comment by danthorpe** [comment]
See #90.

**Comment by danthorpe** [comment]
@Antoine4011 would you be able to test my implantation from #90 for this using this in your Podfile?

```ruby
target 'My iOS Extension' do
    pod 'Operations/Extension', :git => 'https://github.com/danthorpe/Operations', :branch => 'feature/OPR-90_platforms'
end
```

If your extension doesn't compile, can you screenshot/paste the error messages back here please? Thanks!

**Comment by danthorpe** [comment]
Okay, I've tested this myself, and it works as a dependency of say an iOS Today Extension. But with the current version of Cocoapods (0.38.2) it doesn't yet work as a dependency of a watchOS extension - although I think it will with a new Cocoapods.

**Comment by danthorpe** [comment]
Fixed in #90.

**Comment by sivamoorthykutty** [comment]
Hi, I have faced error in ProcedureKit how could I resolve it.
**Error:**
Assertion failed: Calling start() manually on a Procedure is unsupported. Instead, add the Procedure to a ProcedureQueue.: 
/Sources/ProcedureKit/Procedure.swift, line 592
**Error:**
